src/relay/analysis/get_calibration_data.cc

```diff
@@ -0,0 +1,192 @@
+ * steps. First, we need to prepare the module that generate
+ * the tensor values (GetCalibrateModule). Second, we need to
+ * generate the mapping between the values and the subgraphs
+ * (GetCalibrateOutputMap).
+ */
+
+#include <tvm/relay/expr.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/analysis.h>
+
+namespace tvm {
+namespace relay {
+
+/*!
+ * \brief This function returns a module that will be used by
+ * the relay graph runtime for collecting the calibration data.
+ * To do that, we first make all inputs and outputs of each
+ * subgrpah into the final output (i.e., the final output is a
+ * tuple of tensors). Then, we change the compiler attribute of
+ * each subgraph. Finally, we mark all subgraph to be inlined.
+ */
+
+IRModule GetCalibrateModule(IRModule module) {
+  class OutputCollector : public ExprRewriter {
+   public:
+    OutputCollector(const Map<GlobalVar, BaseFunc>& glob_funcs)
+      : glob_funcs(glob_funcs) {}
+
+    Expr Rewrite_(const CallNode* call, const Expr& post) final {
+      if (call->op->IsInstance<GlobalVarNode>()) {
+        auto var = Downcast<GlobalVar>(call->op);
+        // check if it is a subgraph of the original graph
+        if (glob_funcs.count(var) > 0) {
+          for (size_t i = 0; i < call->args.size(); i++)
+            new_outputs.push_back(call->args[i]);
+          // need to flatten the output if it is a tuple
+          auto* fn = glob_funcs[var].as<FunctionNode>();
+          if (auto* tn = fn->body.as<TupleNode>()) {
+            for (size_t i = 0; i < tn->fields.size(); i++) {
+              new_outputs.push_back(TupleGetItem(post, i));
+            }
+          } else {
+            new_outputs.push_back(post);
+          }
+        }
+      }
+      return post;
+    }
+
+    Array<Expr> GetNewOutputs() {
+      return new_outputs;
+    }
+
+   private:
+    const Map<GlobalVar, BaseFunc>& glob_funcs;
+    Array<Expr> new_outputs;
+  };
+
+  auto glob_funcs = module->functions;
+  // module is mutable, hence, we make a copy of it.
+  module.CopyOnWrite();
+  for (const auto& pair : glob_funcs) {
+    if (auto* fn = pair.second.as<FunctionNode>()) {
+      auto func = GetRef<Function>(fn);
+      auto* gl_var = pair.first.as<GlobalVarNode>();
+      // we only collect the outputs for main function
+      if (gl_var->name_hint == "main") {
+        OutputCollector output_collector(glob_funcs);
+        PostOrderRewrite(func->body, &output_collector);
+        auto new_outputs = output_collector.GetNewOutputs();
+        if (!new_outputs.empty()) {
+          Array<Expr> fields;
+          for (const auto& output : new_outputs) {
+            fields.push_back(output);
+          }
+          auto tuple = Tuple(fields);
+          func = Function(func->params, tuple, tuple->checked_type_, 
+                          func->type_params, func->attrs);
+        }
+      // inline the function if it is not main function
```

Comment:
1. Move this comment to the else branch.
2. Add the reason about why we do this.